File: backend/app/schemas/food.py
"""
Food-related Pydantic schemas.
"""
from datetime import date, datetime, time
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, Field


class FoodBase(BaseModel):
    """Base schema for food tracking."""
    meal_date: date
    meal_time: Optional[time] = None
    meal_type: str = Field(..., pattern="^(breakfast|lunch|dinner|snack)$")
    food_name: str = Field(..., max_length=255)
    portion_size: Optional[str] = Field(None, max_length=100)
    calories: Optional[Decimal] = Field(None, ge=0)
    protein_grams: Optional[Decimal] = Field(None, ge=0)
    carbs_grams: Optional[Decimal] = Field(None, ge=0)
    fats_grams: Optional[Decimal] = Field(None, ge=0)
    fiber_grams: Optional[Decimal] = Field(None, ge=0)
    sugar_grams: Optional[Decimal] = Field(None, ge=0)
    sodium_mg: Optional[Decimal] = Field(None, ge=0)
    notes: Optional[str] = None
    is_favorite: bool = False


class FoodCreate(FoodBase):
    """Schema for creating food entry."""
    pass


class FoodUpdate(BaseModel):
    """Schema for updating food entry."""
    meal_time: Optional[time] = None
    food_name: Optional[str] = Field(None, max_length=255)
    portion_size: Optional[str] = Field(None, max_length=100)
    calories: Optional[Decimal] = Field(None, ge=0)
    protein_grams: Optional[Decimal] = Field(None, ge=0)
    carbs_grams: Optional[Decimal] = Field(None, ge=0)
    fats_grams: Optional[Decimal] = Field(None, ge=0)
    fiber_grams: Optional[Decimal] = Field(None, ge=0)
    sugar_grams: Optional[Decimal] = Field(None, ge=0)
    sodium_mg: Optional[Decimal] = Field(None, ge=0)
    notes: Optional[str] = None
    is_favorite: Optional[bool] = None


class FoodResponse(FoodBase):
    """Schema for food response."""
    id: UUID
    user_id: UUID
    created_at: datetime
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True

File: frontend/src/types/food.ts
/**
 * Type definitions for food tracking.
 */

/**
 * Meal type options
 */
export type MealType = 'breakfast' | 'lunch' | 'dinner' | 'snack';

/**
 * Food entry interface
 */
export interface Food {
  id: string;
  user_id: string;
  meal_date: string; // ISO date string
  meal_time?: string; // Time string
  meal_type: MealType;
  food_name: string;
  portion_size?: string;
  calories?: number;
  protein_grams?: number;
  carbs_grams?: number;
  fats_grams?: number;
  fiber_grams?: number;
  sugar_grams?: number;
  sodium_mg?: number;
  notes?: string;
  is_favorite: boolean;
  created_at: string;
  updated_at?: string;
}

/**
 * Request types
 */
export interface CreateFoodRequest {
  meal_date: string;
  meal_time?: string;
  meal_type: MealType;
  food_name: string;
  portion_size?: string;
  calories?: number;
  protein_grams?: number;
  carbs_grams?: number;
  fats_grams?: number;
  fiber_grams?: number;
  sugar_grams?: number;
  sodium_mg?: number;
  notes?: string;
  is_favorite?: boolean;
}

export interface UpdateFoodRequest {
  meal_time?: string;
  food_name?: string;
  portion_size?: string;
  calories?: number;
  protein_grams?: number;
  carbs_grams?: number;
  fats_grams?: number;
  fiber_grams?: number;
  sugar_grams?: number;
  sodium_mg?: number;
  notes?: string;
  is_favorite?: boolean;
}

/**
 * Food statistics
 */
export interface FoodStats {
  total_calories: number;
  total_protein: number;
  total_carbs: number;
  total_fats: number;
  meal_count: number;
}
